File: deck.py
"""
Author: Chow
Create: 2019/05/30
Last Review: 2019/06/01
"""

import random

SUITS = ['spade', 'club', 'heart', 'diamond']

class Deck:
    def __init__(self, duplicate=1, size='no_joker'): 
        self.cards = []
        self.discard_cards = []
        """
        size: full_size, no_joker, num_only, single_suit
        duplicate >= 1
        """
        for _ in range(duplicate): 
            if size == 'no_joker':
                for suit in SUITS:
                    for value in range(1, 14):
                        self.cards.append(Card(value, suit))
        self.shuffle()

    def shuffle(self):
        random.shuffle(self.cards)

    def deal(self, flip=False):
        deal_card = None
        if len(self.cards) > 0:
            deal_card = self.cards.pop(0)
        else:
            self.recycle()
            deal_card = self.cards.pop(0)
        if flip:
            deal_card.flip()
        return deal_card

    def drop(self, card):
        if not card.face_up:
            card.flip()
        self.discard_cards.append(card)

    def recycle(self):
        for card in self.discard_cards:
            card.flip()
        self.cards.extend(self.discard_cards)
        self.discard_cards = []
        self.shuffle()

    def show(self, check=False):
        print('cards: ')
        for card in self.cards:
            card.show(check)
            print(' ', end='')
        print()
        print('discard_cards: ')
        for card in self.discard_cards:
            card.show(check)
            print(' ', end='')
        print()

class Card:
    def __init__(self, value=1, suit=SUITS[0], face_up=False):
        """
        value: 1~15
        suit: spade, club, heart, diamond, None
        """
        self.value = value
        self.suit = suit
        self.face_up = face_up

    def flip(self):
        self.face_up = not self.face_up

    def show(self, check=False):
        if not self.face_up and not check:
            print('Hidden', end='')
        elif self.value == 15:
            print('Joker(Red)', end='')
        elif self.value == 14:
            print('Joker(Black)', end='')
        elif self.value ==13:
            print('{0}-King'.format(self.suit), end='')
        elif self.value ==12:
            print('{0}-Queen'.format(self.suit), end='')
        elif self.value ==11:
            print('{0}-Jack'.format(self.suit), end='')
        else:
            print('{0}-{1}'.format(self.suit, self.value), end='')

if __name__ == "__main__":
    deck = Deck()
    deck.show(True)
    deck.shuffle()
    card = deck.deal()
    deck.drop(card)
    card = deck.deal()
    card.flip()
    deck.drop(card)
    deck.show(False)
    deck.recycle()
    deck.show(False)

File: deck.go
package blackjack

import (
	"fmt"
	"math/rand"
	"time"
)

const (
	spade = iota
	club
	heart
	diamond
	suitNum
)

// deckType
const (
	fullSize = iota
	noJoker
	numOnly
)

type Deck struct {
	cards        []*Card
	discardCards []*Card
}

// """
// size: full_size, no_joker, num_only, single_suit
// duplicate >= 1
// """
func NewDeck(duplicate, size int) *Deck {
	deck := &Deck{}
	deck.cards = make([]*Card, 0)
	deck.discardCards = make([]*Card, 0)
	for dup := 0; dup < duplicate; dup++ {
		if size == noJoker {
			for suit := 0; suit < suitNum; suit++ {
				for value := 1; value < 14; value++ {
					deck.cards = append(deck.cards, NewCard(value, suit, false))
				}
			}
		}
	}
	deck.shuffle()
	return deck
}

func (d *Deck) shuffle() {
	rand.Seed(time.Now().UnixNano())
	rand.Shuffle(len(d.cards), func(i, j int) {
		d.cards[i], d.cards[j] = d.cards[j], d.cards[i]
	})
}

func (d *Deck) deal(flip bool) *Card {
	var card *Card
	if len(d.cards) > 0 {
		card = d.cards[0]
		if len(d.cards) == 1 {
			d.cards = make([]*Card, 0)
		} else {
			d.cards = d.cards[1:]
		}
	} else if len(d.cards) == 0 {
		d.recycle()
		card = d.deal(false)
	}
	if flip {
		card.flip()
	}
	return card
}

func (d *Deck) drop(card *Card) {
	if !card.faceUp {
		card.flip()
	}
	d.discardCards = append(d.discardCards, card)
}

func (d *Deck) recycle() {
	for _, card := range d.discardCards {
		card.flip()
		d.cards = append(d.cards, card)
	}
	d.discardCards = make([]*Card, 0)
	d.shuffle()
}

func (d *Deck) show(check bool) {
	fmt.Print("cards: ")
	for _, card := range d.cards {
		card.show(check)
		fmt.Print(" ")
	}
	fmt.Println()
	fmt.Print("discardCards: ")
	for _, card := range d.discardCards {
		card.show(check)
		fmt.Print(" ")
	}
	fmt.Println()
}

type Card struct {
	value  int
	suit   int
	faceUp bool
}

// """
// value: 1~15
// suit: spade, club, heart, diamond, None
// """
func NewCard(value, suit int, faceUp bool) *Card {
	card := &Card{
		value,
		suit,
		faceUp,
	}
	return card
}

func (c *Card) flip() {
	c.faceUp = !c.faceUp
}

func (c *Card) show(check bool) {
	if !c.faceUp && !check {
		fmt.Print("Hidden")
	} else if c.value == 15 {
		fmt.Print("Joker(Red)")
	} else if c.value == 14 {
		fmt.Print("Joker(Black)")
	} else if c.value == 13 {
		fmt.Print(fmt.Sprintf("%d-King", c.suit))
	} else if c.value == 12 {
		fmt.Print(fmt.Sprintf("%d-Queen", c.suit))
	} else if c.value == 11 {
		fmt.Print(fmt.Sprintf("%d-Jack", c.suit))
	} else {
		fmt.Print(fmt.Sprintf("%d-%d", c.suit, c.value))
	}
}
